packages/jupyter-ai/jupyter_ai/chat_handlers/tools.py

```diff
@@ -0,0 +1,219 @@
+import argparse
+import math
+
+# LangGraph imports for using tools
+import os
+import re
+from typing import Dict, Literal, Type
+
+import numpy as np
+from jupyter_ai.models import HumanChatMessage
+from jupyter_ai_magics.providers import BaseProvider
+from langchain.chains import ConversationalRetrievalChain, LLMChain
+from langchain.memory import ConversationBufferWindowMemory
+from langchain_anthropic import ChatAnthropic
+
+# Chat Providers (add more as needed)
+from langchain_aws import ChatBedrock
+from langchain_cohere import ChatCohere
+from langchain_core.messages import AIMessage
+from langchain_core.prompts import PromptTemplate
+from langchain_core.tools import tool
+from langchain_google_genai import ChatGoogleGenerativeAI
+from langchain_ollama import ChatOllama
+from langchain_openai import AzureChatOpenAI, ChatOpenAI
+from langgraph.graph import MessagesState, StateGraph
+from langgraph.prebuilt import ToolNode
+
+from .base import BaseChatHandler, SlashCommandRoutingType
+
+PROMPT_TEMPLATE = """Given the following conversation and a follow up question, rephrase the follow up question to be a standalone question.
+
+Chat History:
+{chat_history}
+Follow Up Input: {question}
+Standalone question:
+Format the answer to be as pretty as possible.
+"""
+CONDENSE_PROMPT = PromptTemplate.from_template(PROMPT_TEMPLATE)
+
+
+class ToolsChatHandler(BaseChatHandler):
+    """Processes messages prefixed with /tools. This actor will
+    bind a <tool_name>.py collection of tools to the LLM and
+    build a computational graph to direct queries to tools
+    that apply to the prompt. If there is no appropriate tool,
+    the LLM will default to a standard chat response from the LLM
+    without using tools.
+    """
+
+    id = "tools"
+    name = "Use tools with LLM"
+    help = "Ask a question that uses your custom tools"
+    routing_type = SlashCommandRoutingType(slash_id="tools")
+
+    uses_llm = True
+
+    # def __init__(self, retriever, *args, **kwargs):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # self._retriever = retriever
+        self.parser.prog = "/tools"
+        self.parser.add_argument("query", nargs=argparse.REMAINDER)
+        self.tools_file_path = os.path.join(
```

Comment:
Something to think about is the usage case of users updating the tools file while the server is running. Many users use Jupyter in a managed setting where they don't have the ability to restart the server. Because of this, it would be helpful for users to be able to update the file and have the new or revised tools immediately be available. Because of this, I think the file should be read each time the `/tools` slash command is used.